### PR TITLE
update json-patch to fix nil value issue when creating mergepatch

### LINKED ISSUE #49044: kubectl annotate patches too much
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
I used `kubectl --v=10 annotate` to learn about patching object metadata, and was surprised to find more than the metadata in the PATCH request body.  Here is an example:
```
    $ kubectl --v=10 annotate pods mypod bar=foo
    ...
    I0717 18:55:21.106675  161234 request.go:991] Request Body: {"metadata":{"annotations":{"bar":"foo"}},"status":{"conditions":[{"lastProbeTime":null,"lastTransitionTime":"2017-07-14T00:36:40Z","status":"True","type":"Initialized"},{"lastProbeTime":null,"lastTransitionTime":"2017-07-14T00:37:06Z","status":"True","type":"Ready"},{"lastProbeTime":null,"lastTransitionTime":"2017-07-14T00:36:40Z","status":"True","type":"PodScheduled"}]}}

    I0717 18:55:21.106795  161234 round_trippers.go:398] curl -k -v -XPATCH  -H "Content-Type: application/merge-patch+json" -H "User-Agent: kubectl/v1.6.4 (linux/amd64) kubernetes/d6f4332" -H "Accept: application/json" https://172.16.11.12:443/api/v1/namespaces/default/pods/mypod

    I0717 18:55:21.112228  161234 round_trippers.go:417] PATCH https://172.16.11.12:443/api/v1/namespaces/default/pods/mypod 200 OK in 5 milliseconds
```

**What you expected to happen**:
I expected to see my new annotation, but not the pod's status, in the request body.

**How to reproduce it (as minimally and precisely as possible)**:
I hope the above makes this clear.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.6.4 in client and server
- Cloud provider or hardware configuration**: bare metal install
- OS (e.g. from /etc/os-release): Ubuntu 16.04
- Kernel (e.g. `uname -a`): 4.4.0-34-generic
- Install tools: private Ansible
- Others: N/A

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
When [creating a patch for merge](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/annotate.go#L255), nil value will be considered as different value. This has been fixed and merged in [evanphx/json-patch #45](https://github.com/evanphx/json-patch/pull/45).

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #49044

**Special notes for your reviewer**:
/cc @MikeSpreitzer @mengqiy 

**Release note**:

```release-note
Fix nil value issue when creating json patch for merge
```

### CODE REVIEW

**Comment by dixudx** [comment]
@mengqiy Need your approval. Thanks.

**Comment by mengqiy** [comment]
Want to make sure that all the test files change under `pkg/kubectl/cmd/testdata/edit/` are all auto-recorded. You didn't manually change them, right?

**Comment by dixudx** [comment]
@mengqiy I manually change them. Also I don't know why `~1` are kept left there.

**Comment by dixudx** [comment]
@shiywang I noticed test files under `pkg/kubectl/cmd/testdata/edit/` were created in your PR #42256. Does `~1` in `kubectl.kubernetes.io~1last-applied-configuration`  have some special meanings? I manually replace `~1` with `/`. @mengqiy and I want to evaluate the impact. Thanks.

**Comment by dixudx** [comment]
@lavalamp @brendandburns PTAL. Thanks.

File: pkg/kubectl/cmd/testdata/edit/testcase-apply-edit-last-applied-syntax-error/3.request [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
     "metadata": {
         "annotations": {
-            "kubectl.kubernetes.io~1last-applied-configuration": "{\"kind\":\"Service\",\"metadata\":{\"annotations\":{},\"labels\":{\"app\":\"svc1\",\"new-label\":\"foo\",\"new-label1\":\"foo1\"},\"name\":\"svc1\",\"namespace\":\"myproject\"},\"spec\":{\"ports\":[{\"name\":\"80\",\"port\":81,\"protocol\":\"TCP\",\"targetPort\":81}],\"sessionAffinity\":\"None\",\"type\":\"ClusterIP\"},\"status\":{\"loadBalancer\":{}}}\n"
+            "kubectl.kubernetes.io/last-applied-configuration": "{\"kind\":\"Service\",\"metadata\":{\"annotations\":{},\"labels\":{\"app\":\"svc1\",\"new-label\":\"foo\",\"new-label1\":\"foo1\"},\"name\":\"svc1\",\"namespace\":\"myproject\"},\"spec\":{\"ports\":[{\"name\":\"80\",\"port\":81,\"protocol\":\"TCP\",\"targetPort\":81}],\"sessionAffinity\":\"None\",\"type\":\"ClusterIP\"},\"status\":{\"loadBalancer\":{}}}\n"
```
Comment (shiywang) [inline pkg/kubectl/cmd/testdata/edit/testcase-apply-edit-last-applied-syntax-error/3.request]: please don't do any manually editing even it's right, please use `kubectl apply ...  --v=loglevel` to recording request and response, and paste the result.

Comment (mengqiy) [inline pkg/kubectl/cmd/testdata/edit/testcase-apply-edit-last-applied-syntax-error/3.request]: No need to copy and paste.
Just set `UPDATE_EDIT_FIXTURE_DATA` to true, the test will auto record the result.

Comment (mengqiy) [inline pkg/kubectl/cmd/testdata/edit/testcase-apply-edit-last-applied-syntax-error/3.request]: See: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/edit_test.go#L87

Comment (dixudx) [inline pkg/kubectl/cmd/testdata/edit/testcase-apply-edit-last-applied-syntax-error/3.request]: @mengqiy It works. Excellent!

@shiywang The files just updated and are exactly the same with my changes.

I think it is safe to get merged.

**Comment by dixudx** [comment]
@brendandburns @dchen1107 @lavalamp @jbeda @mengqiy Need your `lgtm` and `approval`.

**Pull request merged**